app/actions/offset.ts

```diff
@@ -169,39 +172,62 @@ export const retireCarbonTransaction = async (params: {
 
     // retire transaction
     const retireContract = getContract({
-      contractName: "retirementAggregator",
+      contractName: "retirementAggregatorV2",
       provider: params.provider.getSigner(),
     });
 
     params.onStatus("userConfirmation");
+    let sourceAmount: any;
+
+    const parsed = utils.parseUnits(
+      params.quantity,
+      getTokenDecimals(params.retirementToken)
+    );
 
+    if (params.projectAddress) {
+      sourceAmount = await retireContract.getSourceAmountSpecificRetirement(
+        addresses["mainnet"][params.inputToken],
+        addresses["mainnet"][params.retirementToken],
+        parsed
+      );
+    } else {
+      sourceAmount = await retireContract.getSourceAmountDefaultRetirement(
+        addresses["mainnet"][params.inputToken],
+        addresses["mainnet"][params.retirementToken],
+        parsed
+      );
+    }
     let txn;
     if (!!params.projectAddress) {
-      txn = await retireContract.retireCarbonSpecific(
+      txn = await retireContract.retireExactCarbonSpecific(
         addresses["mainnet"][params.inputToken],
         addresses["mainnet"][params.retirementToken],
+        params.projectAddress,
+        sourceAmount,
```

Comment:
Assuming this parameter, `maxAmountIn` is meant to prevent slippage, i think we should pass in the value that we already approved for spend, which should happen before we invoke this method. I asked cujo/shimon if my reasoning is correct here

Reply:
so just retrieve the allowance for that particular token and use that for `maxAmountIn`? I was working off the docs, which said to use the `getSourceAmount... `functions for the maxAmountIn param:

> Maximum amount of the source token to spend. To find the exact amount needed, please refer to the view functions getSourceAmountSpecificRetirement and getSourceAmountDefaultRetirement